ENH/TST: Branch: 0.10 - firewallcmd action tests

TorontoMedia opened the pull request:
- added ipmset to firewallcmd-common.
- added 3 firewallcmd action tests[ firewallcmd-new,  firewallcmd-rich-rules,  firewallcmd-rich-logging ] in 

>  testCheckStockCommandActions

yarikoptic commented on config/action.d/firewallcmd-common.conf:
```diff
@@ -5,72 +5,102 @@
 
 # Option:  chain
-# Notes    specifies the firewalld chain to which the Fail2Ban rules should be
+# Notes:   Specifies the firewalld chain to which the Fail2Ban rules should be
```
Minor... I guess no need to capitalize since is lower case in others

yarikoptic replied:
It actually it all varies... Might be worth unifying...

TorontoMedia replied:
I have some time today so let me know what you'd like to see regarding format and I'll work on it.

yarikoptic commented on config/action.d/firewallcmd-common.conf:
```diff
@@ -5,72 +5,102 @@
+# Notes:   Specifies the bantime in seconds (handled internally rather than by fail2ban)
+# Values:  [ NUM ]  Default: 600
+
+bantime = 600
```
BTW... I wonder if ideally we should somehow flag (or interface this bantime into fail 2ban) such actions which time out independently of our internal logic. Since eg of you set bantime here shorter than internally known bantime , wouldn't we even not try to ban until internal timer for the ban runs out, @sebres?

yarikoptic replied:
Or are we passing bantime setting into the jails actions and I just forgot about that, and it is a non issue? :-)

sebres replied:
Imho, the `bantime` is only required for the actions, that are not working without it (expected in cmd-line), resp. has no unban-functionality at all (self-unbanning).
For many of these a default timeout (0 or -1) for infinite can be set, so will be unbanned by remove rule during call of unban (except these, that are not have it).
The passing of a bantime is configurable even as `bantime` parameter in brackets by defining of an action in jail.conf.
Why and whether it will be expected in firewallcmd I don't know (I don't use it at all)...

We will get a problem with such actions, as long as we will integrate ban-time-incr branch. Because these actions should be treated differently. 
Roughly described flow that currently implemented in ban-time-incr:
- if failure occurred, it will be added to failmanager immediately, that send an event to observer
- asynchronous observer checks the database, to determine IP would be bad, if yes observer "decreases" `maxretry` for this IP;
- ban occurred immediately if reached `maxretry`;
- asynchronous observer checks the database, to determine IP would be bad;
- if bad, observer prolongs a time of ban in banmanager only (without calling any action again)
- ...
- banmanager calls unban if the ban time was reached

TorontoMedia replied:
Seeing as though an incrementing bantime doesn't matter in firewallcmd-ipset unban or any other firewallcmd at the time being, does anyone see a problem in it being defined in -common?

TorontoMedia replied:
maybe a new action:
> `actionincr = ipset add fail2ban-<name> <ip> timeout <bantime> -exist`

or 

> `actiontimeout = ipset add fail2ban-<name> <ip> timeout <bantime> -exist`

TorontoMedia commented:
@fail2ban/contributors, @fail2ban/developers Can we not change firewallcmd-new cmd state module to conntrack and drop firewallcmd-multiport or just drop firewallcmd-new and their related test?

sebres commented:
About the drop of firewallcmd-new (resp. also iptables-new) from my side - I'll like if we have fewer actions, ideally put all almost doing the same actions together to 1 or 2 general actions (configurable using parameters). We've indeed too many almost identical actions.

There is still the question backwards compatibility, but we can add an example config for each obsolete action for replacement with new general action.

For example: difference between ...-allports and ...-multiport is minimal, and could be completely covered with a sole parameter. 

I can also extend conditional syntax in config readers to make it more dynamically, but currently the problem is only sections can wear the condition, because there are not allowed to use many special characters in name of parameters (`=`, `:`, etc).
But it would be possible with something like:
```
[Init?type=multiport]
actoptions = -m conntrack --ctstate NEW -p <protocol> -m multiport --dports <port>
```

sebres commented:
Another idea without conditional section via `actoptions`:
* firewallcmd.conf:
```
actionstart = firewall-cmd --direct --add-chain <family> filter f2b-<name>
              firewall-cmd --direct --add-rule <family> filter f2b-<name> 1000 -j RETURN
              firewall-cmd --direct --add-rule <family> filter <chain> 0 <actoptions> -j f2b-<name>
...
actoptions =
multiport = -m conntrack --ctstate NEW -p <protocol> -m multiport --dports <port>
allports = 
```
* jail.conf:
```
banaction = firewallcmd[actoptions=<multiport>]
banaction_allports = firewallcmd[actoptions=<allports>]
```

Currently one simple todo to be fixed - putting together N blocks of parameters, because I think, following will be currently not interpreted:
```
action = act[p1=...][p2=...]; # wrong
action = act[p1=... p2=...]; # legal
```

yarikoptic commented:
Regarding multiple -- Aren't they comma separated?

On May 15, 2016 9:22:56 AM EDT, "Serg G. Brester" <notifications@github.com> wrote:
>Another idea without conditional section via `actoptions`:
>* firewallcmd.conf:
>```
>actionstart = firewall-cmd --direct --add-chain <family> filter
>f2b-<name>
>firewall-cmd --direct --add-rule <family> filter f2b-<name> 1000 -j
>RETURN
>firewall-cmd --direct --add-rule <family> filter <chain> 0 <actoptions>
>-j f2b-<name>
>...
>actoptions =
>multiport = -m conntrack --ctstate NEW -p <protocol> -m multiport
>--dports <port>
>allports = 
>```
>* jail.conf:
>```
>banaction = firewallcmd[actoptions=<multiport>]
>banaction_allports = firewallcmd[actoptions=<allports>]
>```
>
>Currently one simple todo to be fixed - putting together N blocks of
>parameters, because I think, following will be currently not
>interpreted:
>```
>action = act[p1=...][p2=...]; # wrong
>action = act[p1=... p2=...]; # legal
>```
>
>---
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/fail2ban/fail2ban/pull/1425#issuecomment-219285524

sebres commented:
> Aren't they comma separated?

fixed

sebres commented:
Because all these parameters are fix (not modified during processing) we can do it much easy. Put it all direct in definition section (not in init) and use another substitution:
```
banaction = firewallcmd[actoptions=%(multiport)s]
banaction_allports = firewallcmd[actoptions=%(allports)s]
```

I've currently too many todo's to make it ready, so will see later...

yarikoptic commented:
But those multiport, allport options wouldn't be universally applicable to all major actions we could use for banning, so not sure if I like it defining here. Fwiw The idea of having those modes defined in the action files sounded better to me. 

On May 15, 2016 9:36:46 AM EDT, "Serg G. Brester" <notifications@github.com> wrote:
>Because all these parameters are fix (not modified during processing)
>we can do it much easy. Put it all direct in definition section (not in
>init) and use another substitution:
>```
>banaction = firewallcmd[actoptions=%(multiport)s]
>banaction_allports = firewallcmd[actoptions=%(allports)s]
>```
>
>I've currently too many todo's to make it ready, so will see later...
>
>---
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/fail2ban/fail2ban/pull/1425#issuecomment-219286242

TorontoMedia commented:
@sebres I like your idea and support the idea of reducing actions but for the time being I agree with yarikoptic in the sense that firewalld actions should use the defined default action. 

But for now we could take your advice and remove the use of the obsolete module(state) and define it in multiport or vise-versa and drop one action...or we could explain that state is obsolete and to use firewallcmd-multiport if you have the conntrack module in firewallcmd-new?

sebres commented:
@TorontoMedia support for multiple option groups is ready in #1431

TorontoMedia commented:
@sebres Okay thanks, I'll update the actions when I get a chance...I just got off working a double shift so I don't have energy to do it right now.

sebres commented:
@TorontoMedia JFYI, please take a look in `pf.conf` from #1429, I mean how it realized there with `actiontype` (just to be consistent in naming)

TorontoMedia commented:
@sebres I'm working long shifts this week and I'm having interpolation errors but I'm too tired to look into it so I'm just going to create this branch: [split-conf](https://github.com/TorontoMedia/fail2ban/tree/split-conf) for now and finish them all Friday.

TorontoMedia commented:
Could someone look at this branch: [split-conf](https://github.com/TorontoMedia/fail2ban/blob/split-conf/config/action.d/firewallcmd-ipset.conf) and explain why bantime is set to 600 but is overwritten with the default 10m please.

> 2016-05-20 11:32:51,152 fail2ban.utils          [11467]: ERROR    -- stderr: "ipset v6.25.1: Syntax error: '10m' is invalid as number"

sebres commented:
will take a look tomorrow, should not be so, this miss str2seconds somewhere

TorontoMedia commented:
@sebres thank you. 

Here's the complete log:

> 2016-05-20 18:09:06,263 fail2ban.server         [16841]: INFO    Start Fail2ban v0.10.0a1
> 2016-05-20 18:09:06,263 fail2ban.server         [16841]: INFO    Changed logging target to /var/log/fail2ban.log for Fail2ban v0.10.0a1
> 2016-05-20 18:09:06,264 fail2ban.transmitter    [16841]: DEBUG   Command: ['set', 'dbfile', '/var/lib/fail2ban/fail2ban.sqlite3']
> 2016-05-20 18:09:06,265 fail2ban.database       [16841]: INFO    Connected to fail2ban persistent database '/var/lib/fail2ban/fail2ban.sqlite3'
> 2016-05-20 18:09:06,330 fail2ban.transmitter    [16841]: DEBUG   Command: ['set', 'dbpurgeage', '1d']
> 2016-05-20 18:09:06,332 fail2ban.transmitter    [16841]: DEBUG   Command: ['add', 'apache-auth', 'auto']
> 2016-05-20 18:09:06,333 fail2ban.jail           [16841]: INFO    Creating new jail 'apache-auth'
> 2016-05-20 18:09:06,336 fail2ban.jail           [16841]: DEBUG   Backend 'pyinotify' failed to initialize due to No module named pyinotify
> 2016-05-20 18:09:06,337 fail2ban.jail           [16841]: DEBUG   Backend 'gamin' failed to initialize due to No module named gamin
> 2016-05-20 18:09:06,338 fail2ban.jail           [16841]: INFO    Jail 'apache-auth' uses poller
> 2016-05-20 18:09:06,339 fail2ban.filter         [16841]: DEBUG   Setting usedns = warn for FilterPoll(Jail('apache-auth'))
> 2016-05-20 18:09:06,341 fail2ban.filter         [16841]: DEBUG   Created FilterPoll(Jail('apache-auth'))
> 2016-05-20 18:09:06,341 fail2ban.filter         [16841]: INFO    Set jail log file encoding to UTF-8
> 2016-05-20 18:09:06,341 fail2ban.filterpoll     [16841]: DEBUG   Created FilterPoll
> 2016-05-20 18:09:06,342 fail2ban.jail           [16841]: INFO    Initiated 'polling' backend
> 2016-05-20 18:09:06,397 fail2ban.transmitter    [16841]: DEBUG   Command: ['set', 'apache-auth', 'usedns', 'warn']
> 2016-05-20 18:09:06,398 fail2ban.filter         [16841]: DEBUG   Setting usedns = warn for FilterPoll(Jail('apache-auth'))
> 2016-05-20 18:09:06,399 fail2ban.transmitter    [16841]: DEBUG   Command: ['set', 'apache-auth', 'addlogpath', '/var/log/apache2/error.log', 'head']
> 2016-05-20 18:09:06,461 fail2ban.filter         [16841]: INFO    Added logfile = /var/log/apache2/error.log (pos = 476, hash = 98f3b018c2dadb674dd765056819554f)
> 2016-05-20 18:09:06,462 fail2ban.transmitter    [16841]: DEBUG   Command: ['set', 'apache-auth', 'maxretry', '5']
> 2016-05-20 18:09:06,463 fail2ban.filter         [16841]: INFO    Set maxRetry = 5
> 2016-05-20 18:09:06,464 fail2ban.transmitter    [16841]: DEBUG   Command: ['set', 'apache-auth', 'addignoreip', '127.0.0.1/8']
> 2016-05-20 18:09:06,465 fail2ban.filter         [16841]: DEBUG   Add 127.0.0.0/8 to ignore list ('127.0.0.1/8')
> 2016-05-20 18:09:06,466 fail2ban.transmitter    [16841]: DEBUG   Command: ['set', 'apache-auth', 'addignoreip', '::1']
> 2016-05-20 18:09:06,467 fail2ban.filter         [16841]: DEBUG   Add ::1 to ignore list ('::1')
> 2016-05-20 18:09:06,468 fail2ban.transmitter    [16841]: DEBUG   Command: ['set', 'apache-auth', 'logencoding', 'auto']
> 2016-05-20 18:09:06,468 fail2ban.filter         [16841]: INFO    Set jail log file encoding to UTF-8
> 2016-05-20 18:09:06,470 fail2ban.transmitter    [16841]: DEBUG   Command: ['set', 'apache-auth', 'bantime', '10m']
> 2016-05-20 18:09:06,470 fail2ban.actions        [16841]: INFO    Set banTime = 600
> 2016-05-20 18:09:06,471 fail2ban.transmitter    [16841]: DEBUG   Command: ['set', 'apache-auth', 'ignorecommand', '']
> 2016-05-20 18:09:06,472 fail2ban.transmitter    [16841]: DEBUG   Command: ['set', 'apache-auth', 'findtime', '10m']
> 2016-05-20 18:09:06,473 fail2ban.filter         [16841]: INFO    Set findtime = 600
> 2016-05-20 18:09:06,474 fail2ban.transmitter    [16841]: DEBUG   Command: ['multi-set', 'apache-auth', 'addfailregex', ['^\\[\\] \\[(:?error|\\S+:\\S+)\\]( \\[pid \\d+(:\\S+ \\d+)?\\])? \\[client <HOST>(:\\d{1,5})?\\] (AH(01797|01630): )?client denied by server configuration: (uri )?\\S*(, referer: \\S+)?\\s*$', '^\\[\\] \\[(:?error|\\S+:\\S+)\\]( \\[pid \\d+(:\\S+ \\d+)?\\])? \\[client <HOST>(:\\d{1,5})?\\] (AH01617: )?user .*? authentication failure for "\\S*": Password Mismatch(, referer: \\S+)?$', '^\\[\\] \\[(:?error|\\S+:\\S+)\\]( \\[pid \\d+(:\\S+ \\d+)?\\])? \\[client <HOST>(:\\d{1,5})?\\] (AH01618: )?user .*? not found(: )?\\S*(, referer: \\S+)?\\s*$', '^\\[\\] \\[(:?error|\\S+:\\S+)\\]( \\[pid \\d+(:\\S+ \\d+)?\\])? \\[client <HOST>(:\\d{1,5})?\\] (AH01614: )?client used wrong authentication scheme: \\S*(, referer: \\S+)?\\s*$', '^\\[\\] \\[(:?error|\\S+:\\S+)\\]( \\[pid \\d+(:\\S+ \\d+)?\\])? \\[client <HOST>(:\\d{1,5})?\\] (AH\\d+: )?Authorization of user \\S+ to access \\S* failed, reason: .*$', '^\\[\\] \\[(:?error|\\S+:\\S+)\\]( \\[pid \\d+(:\\S+ \\d+)?\\])? \\[client <HOST>(:\\d{1,5})?\\] (AH0179[24]: )?(Digest: )?user .*?: password mismatch: \\S*(, referer: \\S+)?\\s*$', "^\\[\\] \\[(:?error|\\S+:\\S+)\\]( \\[pid \\d+(:\\S+ \\d+)?\\])? \\[client <HOST>(:\\d{1,5})?\\] (AH0179[01]: |Digest: )user `.*?' in realm `.+' (not found|denied by provider): \\S*(, referer: \\S+)?\\s*$", '^\\[\\] \\[(:?error|\\S+:\\S+)\\]( \\[pid \\d+(:\\S+ \\d+)?\\])? \\[client <HOST>(:\\d{1,5})?\\] (AH01631: )?user .*?: authorization failure for "\\S*":(, referer: \\S+)?\\s*$', '^\\[\\] \\[(:?error|\\S+:\\S+)\\]( \\[pid \\d+(:\\S+ \\d+)?\\])? \\[client <HOST>(:\\d{1,5})?\\] (AH01775: )?(Digest: )?invalid nonce .* received - length is not \\S+(, referer: \\S+)?\\s*$', "^\\[\\] \\[(:?error|\\S+:\\S+)\\]( \\[pid \\d+(:\\S+ \\d+)?\\])? \\[client <HOST>(:\\d{1,5})?\\] (AH01788: )?(Digest: )?realm mismatch - got `.*?' but expected `.+'(, referer: \\S+)?\\s*$", "^\\[\\] \\[(:?error|\\S+:\\S+)\\]( \\[pid \\d+(:\\S+ \\d+)?\\])? \\[client <HOST>(:\\d{1,5})?\\] (AH01789: )?(Digest: )?unknown algorithm `.*?' received: \\S*(, referer: \\S+)?\\s*$", "^\\[\\] \\[(:?error|\\S+:\\S+)\\]( \\[pid \\d+(:\\S+ \\d+)?\\])? \\[client <HOST>(:\\d{1,5})?\\] (AH01793: )?invalid qop `.*?' received: \\S*(, referer: \\S+)?\\s*$", '^\\[\\] \\[(:?error|\\S+:\\S+)\\]( \\[pid \\d+(:\\S+ \\d+)?\\])? \\[client <HOST>(:\\d{1,5})?\\] (AH01777: )?(Digest: )?invalid nonce .*? received - user attempted time travel(, referer: \\S+)?\\s*$']]
> 2016-05-20 18:09:06,475 fail2ban.server         [16841]: DEBUG     failregex: '^\\[\\] \\[(:?error|\\S+:\\S+)\\]( \\[pid \\d+(:\\S+ \\d+)?\\])? \\[client <HOST>(:\\d{1,5})?\\] (AH(01797|01630): )?client denied by server configuration: (uri )?\\S*(, referer: \\S+)?\\s*$'
> 2016-05-20 18:09:06,479 fail2ban.server         [16841]: DEBUG     failregex: '^\\[\\] \\[(:?error|\\S+:\\S+)\\]( \\[pid \\d+(:\\S+ \\d+)?\\])? \\[client <HOST>(:\\d{1,5})?\\] (AH01617: )?user .*? authentication failure for "\\S*": Password Mismatch(, referer: \\S+)?$'
> 2016-05-20 18:09:06,484 fail2ban.server         [16841]: DEBUG     failregex: '^\\[\\] \\[(:?error|\\S+:\\S+)\\]( \\[pid \\d+(:\\S+ \\d+)?\\])? \\[client <HOST>(:\\d{1,5})?\\] (AH01618: )?user .*? not found(: )?\\S*(, referer: \\S+)?\\s*$'
> 2016-05-20 18:09:06,488 fail2ban.server         [16841]: DEBUG     failregex: '^\\[\\] \\[(:?error|\\S+:\\S+)\\]( \\[pid \\d+(:\\S+ \\d+)?\\])? \\[client <HOST>(:\\d{1,5})?\\] (AH01614: )?client used wrong authentication scheme: \\S*(, referer: \\S+)?\\s*$'
> 2016-05-20 18:09:06,492 fail2ban.server         [16841]: DEBUG     failregex: '^\\[\\] \\[(:?error|\\S+:\\S+)\\]( \\[pid \\d+(:\\S+ \\d+)?\\])? \\[client <HOST>(:\\d{1,5})?\\] (AH\\d+: )?Authorization of user \\S+ to access \\S* failed, reason: .*$'
> 2016-05-20 18:09:06,496 fail2ban.server         [16841]: DEBUG     failregex: '^\\[\\] \\[(:?error|\\S+:\\S+)\\]( \\[pid \\d+(:\\S+ \\d+)?\\])? \\[client <HOST>(:\\d{1,5})?\\] (AH0179[24]: )?(Digest: )?user .*?: password mismatch: \\S*(, referer: \\S+)?\\s*$'
> 2016-05-20 18:09:06,500 fail2ban.server         [16841]: DEBUG     failregex: "^\\[\\] \\[(:?error|\\S+:\\S+)\\]( \\[pid \\d+(:\\S+ \\d+)?\\])? \\[client <HOST>(:\\d{1,5})?\\] (AH0179[01]: |Digest: )user `.*?' in realm `.+' (not found|denied by provider): \\S*(, referer: \\S+)?\\s*$"
> 2016-05-20 18:09:06,503 fail2ban.server         [16841]: DEBUG     failregex: '^\\[\\] \\[(:?error|\\S+:\\S+)\\]( \\[pid \\d+(:\\S+ \\d+)?\\])? \\[client <HOST>(:\\d{1,5})?\\] (AH01631: )?user .*?: authorization failure for "\\S*":(, referer: \\S+)?\\s*$'
> 2016-05-20 18:09:06,507 fail2ban.server         [16841]: DEBUG     failregex: '^\\[\\] \\[(:?error|\\S+:\\S+)\\]( \\[pid \\d+(:\\S+ \\d+)?\\])? \\[client <HOST>(:\\d{1,5})?\\] (AH01775: )?(Digest: )?invalid nonce .* received - length is not \\S+(, referer: \\S+)?\\s*$'
> 2016-05-20 18:09:06,510 fail2ban.server         [16841]: DEBUG     failregex: "^\\[\\] \\[(:?error|\\S+:\\S+)\\]( \\[pid \\d+(:\\S+ \\d+)?\\])? \\[client <HOST>(:\\d{1,5})?\\] (AH01788: )?(Digest: )?realm mismatch - got `.*?' but expected `.+'(, referer: \\S+)?\\s*$"
> 2016-05-20 18:09:06,514 fail2ban.server         [16841]: DEBUG     failregex: "^\\[\\] \\[(:?error|\\S+:\\S+)\\]( \\[pid \\d+(:\\S+ \\d+)?\\])? \\[client <HOST>(:\\d{1,5})?\\] (AH01789: )?(Digest: )?unknown algorithm `.*?' received: \\S*(, referer: \\S+)?\\s*$"
> 2016-05-20 18:09:06,517 fail2ban.server         [16841]: DEBUG     failregex: "^\\[\\] \\[(:?error|\\S+:\\S+)\\]( \\[pid \\d+(:\\S+ \\d+)?\\])? \\[client <HOST>(:\\d{1,5})?\\] (AH01793: )?invalid qop `.*?' received: \\S*(, referer: \\S+)?\\s*$"
> 2016-05-20 18:09:06,520 fail2ban.server         [16841]: DEBUG     failregex: '^\\[\\] \\[(:?error|\\S+:\\S+)\\]( \\[pid \\d+(:\\S+ \\d+)?\\])? \\[client <HOST>(:\\d{1,5})?\\] (AH01777: )?(Digest: )?invalid nonce .*? received - user attempted time travel(, referer: \\S+)?\\s*$'
> 2016-05-20 18:09:06,524 fail2ban.transmitter    [16841]: DEBUG   Command: ['set', 'apache-auth', 'addaction', 'firewallcmd-ipset']
> 2016-05-20 18:09:06,524 fail2ban.CommandAction  [16841]: DEBUG   Created <class 'fail2ban.server.action.CommandAction'>
> 2016-05-20 18:09:06,528 fail2ban.transmitter    [16841]: DEBUG   Command: ['multi-set', 'apache-auth', 'action', 'firewallcmd-ipset', [['actionstop', 'firewall-cmd --direct --remove-rule <family> filter <chain> 0 -p <protocol> -m multiport --dports <port> -m set --match-set <ipmset> src -j <blocktype>\nipset flush <ipmset>\nipset destroy <ipmset>'], ['actioncheck', 'firewall-cmd --direct --query-rule <family> filter <chain> 0 -p <protocol> -m multiport --dports <port> -m set --match-set <ipmset> src -j <blocktype>'], ['actionban', 'ipset add <ipmset> <ip> timeout <bantime> -exist'], ['actionstart', 'ipset create <ipmset> hash:ip timeout <bantime> family <ipmsetfamily>\nfirewall-cmd --direct --add-rule <family> filter <chain> 0 -p <protocol> -m multiport --dports <port> -m set --match-set <ipmset> src -j <blocktype>'], ['actionunban', 'ipset del <ipmset> <ip> -exist'], ['known/ipmset', 'f2b-<name>'], ['protocol', 'tcp'], ['family', 'ipv4'], ['known/rejecttype?family=inet6', 'icmp6-port-unreachable'], ['ipmset', 'f2b-<name>'], ['known/family', 'ipv4'], ['known/family?family=inet6', 'ipv6'], ['ipmsetfamily?family=inet6', 'inet6'], ['chain', 'INPUT'], ['service', 'ssh'], ['known/protocol', 'tcp'], ['port', 'http,https'], ['known/ipmset?family=inet6', 'f2b-<name>6'], ['known/name', 'default'], ['known/service', 'ssh'], ['zone', 'public'], ['known/bantime', '600'], ['bantime', '10m'], ['known/zone', 'public'], ['known/rejecttype', 'icmp-port-unreachable'], ['known/ipmsetfamily', 'inet'], ['rich-blocktype', "reject type='<rejecttype>'"], ['ipmsetfamily', 'inet'], ['rejecttype', 'icmp-port-unreachable'], ['rejecttype?family=inet6', 'icmp6-port-unreachable'], ['blocktype', 'REJECT --reject-with <rejecttype>'], ['ipmset?family=inet6', 'f2b-<name>6'], ['name', 'apache-auth'], ['known/port', '1:65535'], ['family?family=inet6', 'ipv6'], ['known/chain', 'INPUT_direct'], ['known/blocktype', 'REJECT --reject-with <rejecttype>'], ['known/rich-blocktype', "reject type='<rejecttype>'"], ['known/ipmsetfamily?family=inet6', 'inet6']]]
> 2016-05-20 18:09:06,528 fail2ban.transmitter    [16841]: DEBUG     ['actionstop', 'firewall-cmd --direct --remove-rule <family> filter <chain> 0 -p <protocol> -m multiport --dports <port> -m set --match-set <ipmset> src -j <blocktype>\nipset flush <ipmset>\nipset destroy <ipmset>']
> 2016-05-20 18:09:06,528 fail2ban.CommandAction  [16841]: DEBUG     Set actionstop = 'firewall-cmd --direct --remove-rule <family> filter <chain> 0 -p <protocol> -m multiport --dports <port> -m set --match-set <ipmset> src -j <blocktype>\nipset flush <ipmset>\nipset destroy <ipmset>'
> 2016-05-20 18:09:06,528 fail2ban.transmitter    [16841]: DEBUG     ['actioncheck', 'firewall-cmd --direct --query-rule <family> filter <chain> 0 -p <protocol> -m multiport --dports <port> -m set --match-set <ipmset> src -j <blocktype>']
> 2016-05-20 18:09:06,528 fail2ban.CommandAction  [16841]: DEBUG     Set actioncheck = 'firewall-cmd --direct --query-rule <family> filter <chain> 0 -p <protocol> -m multiport --dports <port> -m set --match-set <ipmset> src -j <blocktype>'
> 2016-05-20 18:09:06,528 fail2ban.transmitter    [16841]: DEBUG     ['actionban', 'ipset add <ipmset> <ip> timeout <bantime> -exist']
> 2016-05-20 18:09:06,529 fail2ban.CommandAction  [16841]: DEBUG     Set actionban = 'ipset add <ipmset> <ip> timeout <bantime> -exist'
> 2016-05-20 18:09:06,529 fail2ban.transmitter    [16841]: DEBUG     ['actionstart', 'ipset create <ipmset> hash:ip timeout <bantime> family <ipmsetfamily>\nfirewall-cmd --direct --add-rule <family> filter <chain> 0 -p <protocol> -m multiport --dports <port> -m set --match-set <ipmset> src -j <blocktype>']
> 2016-05-20 18:09:06,529 fail2ban.CommandAction  [16841]: DEBUG     Set actionstart = 'ipset create <ipmset> hash:ip timeout <bantime> family <ipmsetfamily>\nfirewall-cmd --direct --add-rule <family> filter <chain> 0 -p <protocol> -m multiport --dports <port> -m set --match-set <ipmset> src -j <blocktype>'
> 2016-05-20 18:09:06,529 fail2ban.transmitter    [16841]: DEBUG     ['actionunban', 'ipset del <ipmset> <ip> -exist']
> 2016-05-20 18:09:06,529 fail2ban.CommandAction  [16841]: DEBUG     Set actionunban = 'ipset del <ipmset> <ip> -exist'
> 2016-05-20 18:09:06,529 fail2ban.transmitter    [16841]: DEBUG     ['known/ipmset', 'f2b-<name>']
> 2016-05-20 18:09:06,529 fail2ban.CommandAction  [16841]: DEBUG     Set known/ipmset = 'f2b-<name>'
> 2016-05-20 18:09:06,529 fail2ban.transmitter    [16841]: DEBUG     ['protocol', 'tcp']
> 2016-05-20 18:09:06,530 fail2ban.CommandAction  [16841]: DEBUG     Set protocol = 'tcp'
> 2016-05-20 18:09:06,530 fail2ban.transmitter    [16841]: DEBUG     ['family', 'ipv4']
> 2016-05-20 18:09:06,530 fail2ban.CommandAction  [16841]: DEBUG     Set family = 'ipv4'
> 2016-05-20 18:09:06,530 fail2ban.transmitter    [16841]: DEBUG     ['known/rejecttype?family=inet6', 'icmp6-port-unreachable']
> 2016-05-20 18:09:06,530 fail2ban.CommandAction  [16841]: DEBUG     Set known/rejecttype?family=inet6 = 'icmp6-port-unreachable'
> 2016-05-20 18:09:06,530 fail2ban.transmitter    [16841]: DEBUG     ['ipmset', 'f2b-<name>']
> 2016-05-20 18:09:06,530 fail2ban.CommandAction  [16841]: DEBUG     Set ipmset = 'f2b-<name>'
> 2016-05-20 18:09:06,530 fail2ban.transmitter    [16841]: DEBUG     ['known/family', 'ipv4']
> 2016-05-20 18:09:06,531 fail2ban.CommandAction  [16841]: DEBUG     Set known/family = 'ipv4'
> 2016-05-20 18:09:06,531 fail2ban.transmitter    [16841]: DEBUG     ['known/family?family=inet6', 'ipv6']
> 2016-05-20 18:09:06,531 fail2ban.CommandAction  [16841]: DEBUG     Set known/family?family=inet6 = 'ipv6'
> 2016-05-20 18:09:06,531 fail2ban.transmitter    [16841]: DEBUG     ['ipmsetfamily?family=inet6', 'inet6']
> 2016-05-20 18:09:06,531 fail2ban.CommandAction  [16841]: DEBUG     Set ipmsetfamily?family=inet6 = 'inet6'
> 2016-05-20 18:09:06,531 fail2ban.transmitter    [16841]: DEBUG     ['chain', 'INPUT']
> 2016-05-20 18:09:06,531 fail2ban.CommandAction  [16841]: DEBUG     Set chain = 'INPUT'
> 2016-05-20 18:09:06,531 fail2ban.transmitter    [16841]: DEBUG     ['service', 'ssh']
> 2016-05-20 18:09:06,531 fail2ban.CommandAction  [16841]: DEBUG     Set service = 'ssh'
> 2016-05-20 18:09:06,532 fail2ban.transmitter    [16841]: DEBUG     ['known/protocol', 'tcp']
> 2016-05-20 18:09:06,532 fail2ban.CommandAction  [16841]: DEBUG     Set known/protocol = 'tcp'
> 2016-05-20 18:09:06,532 fail2ban.transmitter    [16841]: DEBUG     ['port', 'http,https']
> 2016-05-20 18:09:06,532 fail2ban.CommandAction  [16841]: DEBUG     Set port = 'http,https'
> 2016-05-20 18:09:06,532 fail2ban.transmitter    [16841]: DEBUG     ['known/ipmset?family=inet6', 'f2b-<name>6']
> 2016-05-20 18:09:06,532 fail2ban.CommandAction  [16841]: DEBUG     Set known/ipmset?family=inet6 = 'f2b-<name>6'
> 2016-05-20 18:09:06,532 fail2ban.transmitter    [16841]: DEBUG     ['known/name', 'default']
> 2016-05-20 18:09:06,532 fail2ban.CommandAction  [16841]: DEBUG     Set known/name = 'default'
> 2016-05-20 18:09:06,533 fail2ban.transmitter    [16841]: DEBUG     ['known/service', 'ssh']
> 2016-05-20 18:09:06,533 fail2ban.CommandAction  [16841]: DEBUG     Set known/service = 'ssh'
> 2016-05-20 18:09:06,533 fail2ban.transmitter    [16841]: DEBUG     ['zone', 'public']
> 2016-05-20 18:09:06,533 fail2ban.CommandAction  [16841]: DEBUG     Set zone = 'public'
> 2016-05-20 18:09:06,533 fail2ban.transmitter    [16841]: DEBUG     ['known/bantime', '600']
> 2016-05-20 18:09:06,533 fail2ban.CommandAction  [16841]: DEBUG     Set known/bantime = '600'
> 2016-05-20 18:09:06,533 fail2ban.transmitter    [16841]: DEBUG     ['bantime', '10m']
> 2016-05-20 18:09:06,533 fail2ban.CommandAction  [16841]: DEBUG     Set bantime = '10m'
> 2016-05-20 18:09:06,534 fail2ban.transmitter    [16841]: DEBUG     ['known/zone', 'public']
> 2016-05-20 18:09:06,534 fail2ban.CommandAction  [16841]: DEBUG     Set known/zone = 'public'
> 2016-05-20 18:09:06,534 fail2ban.transmitter    [16841]: DEBUG     ['known/rejecttype', 'icmp-port-unreachable']
> 2016-05-20 18:09:06,534 fail2ban.CommandAction  [16841]: DEBUG     Set known/rejecttype = 'icmp-port-unreachable'
> 2016-05-20 18:09:06,534 fail2ban.transmitter    [16841]: DEBUG     ['known/ipmsetfamily', 'inet']
> 2016-05-20 18:09:06,534 fail2ban.CommandAction  [16841]: DEBUG     Set known/ipmsetfamily = 'inet'
> 2016-05-20 18:09:06,534 fail2ban.transmitter    [16841]: DEBUG     ['rich-blocktype', "reject type='<rejecttype>'"]
> 2016-05-20 18:09:06,534 fail2ban.CommandAction  [16841]: DEBUG     Set rich-blocktype = "reject type='<rejecttype>'"
> 2016-05-20 18:09:06,534 fail2ban.transmitter    [16841]: DEBUG     ['ipmsetfamily', 'inet']
> 2016-05-20 18:09:06,535 fail2ban.CommandAction  [16841]: DEBUG     Set ipmsetfamily = 'inet'
> 2016-05-20 18:09:06,535 fail2ban.transmitter    [16841]: DEBUG     ['rejecttype', 'icmp-port-unreachable']
> 2016-05-20 18:09:06,535 fail2ban.CommandAction  [16841]: DEBUG     Set rejecttype = 'icmp-port-unreachable'
> 2016-05-20 18:09:06,535 fail2ban.transmitter    [16841]: DEBUG     ['rejecttype?family=inet6', 'icmp6-port-unreachable']
> 2016-05-20 18:09:06,535 fail2ban.CommandAction  [16841]: DEBUG     Set rejecttype?family=inet6 = 'icmp6-port-unreachable'
> 2016-05-20 18:09:06,535 fail2ban.transmitter    [16841]: DEBUG     ['blocktype', 'REJECT --reject-with <rejecttype>']
> 2016-05-20 18:09:06,535 fail2ban.CommandAction  [16841]: DEBUG     Set blocktype = 'REJECT --reject-with <rejecttype>'
> 2016-05-20 18:09:06,535 fail2ban.transmitter    [16841]: DEBUG     ['ipmset?family=inet6', 'f2b-<name>6']
> 2016-05-20 18:09:06,535 fail2ban.CommandAction  [16841]: DEBUG     Set ipmset?family=inet6 = 'f2b-<name>6'
> 2016-05-20 18:09:06,536 fail2ban.transmitter    [16841]: DEBUG     ['name', 'apache-auth']
> 2016-05-20 18:09:06,536 fail2ban.CommandAction  [16841]: DEBUG     Set name = 'apache-auth'
> 2016-05-20 18:09:06,536 fail2ban.transmitter    [16841]: DEBUG     ['known/port', '1:65535']
> 2016-05-20 18:09:06,536 fail2ban.CommandAction  [16841]: DEBUG     Set known/port = '1:65535'
> 2016-05-20 18:09:06,536 fail2ban.transmitter    [16841]: DEBUG     ['family?family=inet6', 'ipv6']
> 2016-05-20 18:09:06,536 fail2ban.CommandAction  [16841]: DEBUG     Set family?family=inet6 = 'ipv6'
> 2016-05-20 18:09:06,536 fail2ban.transmitter    [16841]: DEBUG     ['known/chain', 'INPUT_direct']
> 2016-05-20 18:09:06,536 fail2ban.CommandAction  [16841]: DEBUG     Set known/chain = 'INPUT_direct'
> 2016-05-20 18:09:06,536 fail2ban.transmitter    [16841]: DEBUG     ['known/blocktype', 'REJECT --reject-with <rejecttype>']
> 2016-05-20 18:09:06,537 fail2ban.CommandAction  [16841]: DEBUG     Set known/blocktype = 'REJECT --reject-with <rejecttype>'
> 2016-05-20 18:09:06,537 fail2ban.transmitter    [16841]: DEBUG     ['known/rich-blocktype', "reject type='<rejecttype>'"]
> 2016-05-20 18:09:06,537 fail2ban.CommandAction  [16841]: DEBUG     Set known/rich-blocktype = "reject type='<rejecttype>'"
> 2016-05-20 18:09:06,537 fail2ban.transmitter    [16841]: DEBUG     ['known/ipmsetfamily?family=inet6', 'inet6']
> 2016-05-20 18:09:06,537 fail2ban.CommandAction  [16841]: DEBUG     Set known/ipmsetfamily?family=inet6 = 'inet6'
> 2016-05-20 18:09:06,538 fail2ban.transmitter    [16841]: DEBUG   Command: ['start', 'apache-auth']
> 2016-05-20 18:09:06,538 fail2ban.filterpoll     [16841]: DEBUG   /var/log/apache2/error.log has been modified
> 2016-05-20 18:09:06,539 fail2ban.filter         [16841]: DEBUG   Seek to find time 1463781546.46 (2016-05-20 17:59:06), file size 476
> 2016-05-20 18:09:06,539 fail2ban.filter         [16841]: DEBUG   Position -1 from 476, found time None () within 0 seeks
> 2016-05-20 18:09:06,542 fail2ban.action         [16841]: DEBUG   ipset create f2b-apache-auth hash:ip timeout 10m family inet
> firewall-cmd --direct --add-rule ipv4 filter INPUT 0 -p tcp -m multiport --dports http,https -m set --match-set f2b-apache-auth src -j REJECT --reject-with icmp-port-unreachable
> 2016-05-20 18:09:06,618 fail2ban.jail           [16841]: INFO    Jail 'apache-auth' started
> 2016-05-20 18:09:06,788 fail2ban.utils          [16841]: ERROR   ipset create f2b-apache-auth hash:ip timeout 10m family inet
> firewall-cmd --direct --add-rule ipv4 filter INPUT 0 -p tcp -m multiport --dports http,https -m set --match-set f2b-apache-auth src -j REJECT --reject-with icmp-port-unreachable -- stdout:
> 2016-05-20 18:09:06,788 fail2ban.utils          [16841]: ERROR    -- stdout: "\x1b[91mError: COMMAND_FAILED: '/sbin/iptables -w -t filter -I INPUT_direct 1 -p tcp -m multiport --dports http,https -m set --match-set f2b-apache-auth src -j REJECT --reject-with icmp-port-unreachable' failed: iptables v1.4.21: Set f2b-apache-auth doesn't exist."
> 2016-05-20 18:09:06,788 fail2ban.utils          [16841]: ERROR    -- stdout: ''
> 2016-05-20 18:09:06,789 fail2ban.utils          [16841]: ERROR    -- stdout: "Try `iptables -h' or 'iptables --help' for more information.\x1b[00m"
> 2016-05-20 18:09:06,789 fail2ban.utils          [16841]: ERROR   ipset create f2b-apache-auth hash:ip timeout 10m family inet
> firewall-cmd --direct --add-rule ipv4 filter INPUT 0 -p tcp -m multiport --dports http,https -m set --match-set f2b-apache-auth src -j REJECT --reject-with icmp-port-unreachable -- stderr:
> 2016-05-20 18:09:06,789 fail2ban.utils          [16841]: ERROR    -- stderr: "ipset v6.25.1: Syntax error: '10m' is invalid as number"
> 2016-05-20 18:09:06,789 fail2ban.utils          [16841]: ERROR   ipset create f2b-apache-auth hash:ip timeout 10m family inet
> firewall-cmd --direct --add-rule ipv4 filter INPUT 0 -p tcp -m multiport --dports http,https -m set --match-set f2b-apache-auth src -j REJECT --reject-with icmp-port-unreachable -- returned 13
> 2016-05-20 18:09:06,789 fail2ban.action         [16841]: DEBUG   ipset create f2b-apache-auth6 hash:ip timeout 10m family inet6
> firewall-cmd --direct --add-rule ipv6 filter INPUT 0 -p tcp -m multiport --dports http,https -m set --match-set f2b-apache-auth6 src -j REJECT --reject-with icmp6-port-unreachable
> 2016-05-20 18:09:07,025 fail2ban.utils          [16841]: ERROR   ipset create f2b-apache-auth6 hash:ip timeout 10m family inet6
> firewall-cmd --direct --add-rule ipv6 filter INPUT 0 -p tcp -m multiport --dports http,https -m set --match-set f2b-apache-auth6 src -j REJECT --reject-with icmp6-port-unreachable -- stdout:
> 2016-05-20 18:09:07,025 fail2ban.utils          [16841]: ERROR    -- stdout: "\x1b[91mError: COMMAND_FAILED: '/sbin/ip6tables -w -t filter -I INPUT_direct 1 -p tcp -m multiport --dports http,https -m set --match-set f2b-apache-auth6 src -j REJECT --reject-with icmp6-port-unreachable' failed: ip6tables v1.4.21: Set f2b-apache-auth6 doesn't exist."
> 2016-05-20 18:09:07,025 fail2ban.utils          [16841]: ERROR    -- stdout: ''
> 2016-05-20 18:09:07,025 fail2ban.utils          [16841]: ERROR    -- stdout: "Try `ip6tables -h' or 'ip6tables --help' for more information.\x1b[00m"
> 2016-05-20 18:09:07,025 fail2ban.utils          [16841]: ERROR   ipset create f2b-apache-auth6 hash:ip timeout 10m family inet6
> firewall-cmd --direct --add-rule ipv6 filter INPUT 0 -p tcp -m multiport --dports http,https -m set --match-set f2b-apache-auth6 src -j REJECT --reject-with icmp6-port-unreachable -- stderr:
> 2016-05-20 18:09:07,026 fail2ban.utils          [16841]: ERROR    -- stderr: "ipset v6.25.1: Syntax error: '10m' is invalid as number"
> 2016-05-20 18:09:07,026 fail2ban.utils          [16841]: ERROR   ipset create f2b-apache-auth6 hash:ip timeout 10m family inet6
> firewall-cmd --direct --add-rule ipv6 filter INPUT 0 -p tcp -m multiport --dports http,https -m set --match-set f2b-apache-auth6 src -j REJECT --reject-with icmp6-port-unreachable -- returned 13
> 2016-05-20 18:09:07,026 fail2ban.actions        [16841]: ERROR   Failed to start jail 'apache-auth' action 'firewallcmd-ipset': Error starting action Jail('apache-auth')/firewallcmd-ipset

sebres commented:
@TorontoMedia fixed in 3d99833570d9768c9c2d050fa143f5cae38d7dd4, 
just rebase your branch over it (sorry for the conflicts))), should then work

TorontoMedia commented:
@sebres no problem. I can confirm it works now so when I complete my branch I'll update my pull-request with the new actions and tests.

TorontoMedia commented:
I've compressed firewallcmd-[allports | new | multiport] into one action file and also added logging to all actions in that file. Does anyone think that this action is too nested? [firewallcmd](https://github.com/TorontoMedia/fail2ban/blob/nested/config/action.d/firewallcmd.conf)

I had it reduced to one file for all firewallcmd actions but I thought it was too nested so I split them up.

sebres commented:
Looks good.
Do you want to push it as origin:firewallcmd to replace this branch for this PR?

TorontoMedia commented:
Thanks, it's not complete yet. I still have to finish firewallcmd, the rest of the actions I want to work on and figure out how to retain credit for authorship to the original actions but when thats complete I'll update this pull.
